libzecale/circuits/pairing/bls12_377_pairing.tcc

```diff
@@ -105,420 +194,484 @@ template<typename ppT>
 bls12_377_ate_dbl_gadget<ppT>::bls12_377_ate_dbl_gadget(
     libsnark::protoboard<FqT> &pb,
     const bls12_377_G2_proj<ppT> &R,
+    const bls12_377_G2_proj<ppT> &out_R,
+    const bls12_377_ate_ell_coeffs<ppT> &coeffs,
     const std::string &annotation_prefix)
     : libsnark::gadget<libff::Fr<ppT>>(pb, annotation_prefix)
-    , in_R(R)
-    , out_R(pb, " R")
+    , _in_R(R)
+    , _out_R(out_R)
+    , _out_coeffs(coeffs)
 
     // A = Rx * Ry / 2
-    , A(pb, FMT(annotation_prefix, " A"))
-    , check_A(
-          pb, in_R.X, in_R.Y, A * FqT(2), FMT(annotation_prefix, " check_A"))
+    , _A(pb,
```

Comment:
Let's be consistent on how we build these gadgets here. I most parts of the code base, smart ptrs are used and passed as args before being `reset/new`'ed in the parent gadget's constructor. This contrasts with the existing code.

Reply:
See: https://github.com/clearmatics/zecale/issues/23